### DISCUSSION TITLE
Question on Selector in Go client

### DISCUSSION THREAD

**Discussion by iamrodrigo** [uber/cadence#4421] [Q&A] [answered]
I'm reading the signals document https://cadenceworkflow.io/docs/go-client/signals/#signalwithstart and I'm wondering what's the purpose of `s.Select(ctx)`? Do I need to create a new selector every time I want to create a channel?

**Comment by longquanzheng** [comment, marked as answer]
Selector is to repalce native go select by workflow code for determinism.
 Use workflow.NewSelector(ctx) method to create a Selector instance.
 The interface is to simulate Golang's Select statement.
 For example, the logic of Golang code like below
```go
 	chA := make(chan int)
	chB := make(chan int)
	counter := 0
	for {
		select {
		case i, ok := <- chA:
			if ok{
				counter += i
			}
		case i, ok := <- chB:
			if ok{
				counter += i
			}
		}
	}
```
 should be written as
```go
 	s := workflow.NewSelector(ctx)
	counter := 0
	s.AddReceive(workflow.GetSignalChannel(ctx, "channelA"), func(c workflow.Channel, ok bool) {
		if ok{
			var i int
			c.Receive(ctx, &i)
			counter += i
		}
	})
	s.AddReceive(workflow.GetSignalChannel(ctx, "channelB"), func(c workflow.Channel, ok bool) {
		if ok{
			var i int
			c.Receive(ctx, &i)
			counter += i
		}
	})

	for {
		s.Select(ctx)
	}
```

I am adding some documentation to the Selector API:
```go
	Selector interface {
		// AddReceive adds a ReceiveChannel to the selector. f is invoked when the channel has data or closed.
		// ok == false indicates the channel is closed
		AddReceive(c Channel, f func(c Channel, ok bool)) Selector
		// AddSend adds a SendChannel to the selector. f is invoke when the channel is available to send
		AddSend(c Channel, v interface{}, f func()) Selector
		// AddFuture adds a Future to the selector f is invoked when future is ready
		AddFuture(future Future, f func(f Future)) Selector
		// AddDefault adds a default branch to the selector.
		// f is invoked when non of the other conditions(ReceiveChannel, SendChannel and Future) is met for one call of Select
		AddDefault(f func())
		// Select waits for one of the added conditions to be met and invoke the callback as described above.
		// When none of the added condition is met:
		// 		if there is no Default(added by AddDefault) and , then it will block the current goroutine
		// 		if Default(added by AddDefault) is used, when Default callback will be executed without blocking
		// When more than one of added conditions are met, only one of them will be invoked.
		// Usually it's recommended to use a for loop to drain all of them, and use AddDefault to break out the
		// loop properly(e.g. not missing any received data in channels)
		Select(ctx Context)
	}
```
https://github.com/uber-go/cadence-client/pull/1115
@meiliang86 @Groxx Correct me if I am wrong.

**Comment by longquanzheng** [comment]
>Do I need to create a new selector every time I want to create a channel

You don't have to. You will only need selector if you need a stream from channel. If you only need to receive one message, you can simply use `Receive` or `ReceiveAsync`. 

Selector is useful not only for Signal Channel, but also a combination of other stuff including regular chanel(inter gorotines(mush be started by workflow.Go)), timer, activity/childWF futures etc.

**Reply by Groxx** [reply]
The only place where this makes a difference is when you add a Future:
```go
s := workflow.NewSelector(ctx)
s.AddFuture(f, func(f workflow.Future) {
  // ...
})

// the first select will call ^ the callback whenever the future has a value or an error.
// this happens immediately (non-blocking / in the same decision task) if the future is _already_ complete,
// or at some future time.
s.Select(ctx)

// a second call will block forever though - futures are one-shot _per selector_
s.Select(ctx)
```
So if you make a new selector + add a the same future every time, it'll _re-fire_ that callback every time, because the new selector sees a newly-completed future (in that it's newly added and it is already complete), and calls its callback.  It's not too hard to cause a "hot" infinite loop this way, though thankfully it's usually spotted pretty early.

**Comment by iamrodrigo** [comment]
Epic answer @Groxx @longquanzheng , thanks for that!

**Comment by meiliang86** [comment]
@Groxx recently added a lot of documentation around the Cadence selector. Please check them out and let us know if you still have questions.
